### ISSUE TITLE
Adoption of functional programming in the Dart language

### ISSUE THREAD

**Issue by Ing-Brayan-Martinez** [dart-lang/language#3907]
Greetings, I want to open this problem to use it as a collection of problems, to present various characteristics of the functional programming paradigm. From my point of view I believe that the Dart language is a new generation language because it is fulfilling most of the following characteristics:

1. Have a framework (Flutter) that allows you to create UI or graphical computing through the composition of objects, which will allow you to interact with a graphics engine and also with the GPU
2. Having the ability to compile to multiple platforms (Android, IOS, Linux, Windows, MacOS, Web) and multiple processor architectures (X86, ARM, RISC-V) is very important.
3. Have a simple and comfortable syntax that is very useful to simplify algorithms
4. Have the ability to interoperate with low-level languages ​​such as C++, Rust, Objective-C, this allows you to interact directly with hardware such as CPU, GPU, and NPU

The most important feature that is absent in the language is without a doubt the support for functional programming completely throughout the language, something that most languages ​​in 2024 already have. Functional programming allows the Dart language to be more efficient in the following tasks:

- Improve the implementation of mathematical algorithms, since the syntax that allows functional programming allows any mathematical model to be better represented, this is ideal for applications for scientific use, artificial intelligence and any application that requires an advanced mathematical representation
- Improve the consistency and robustness of the language to avoid Bugs within the base SDK, and reduce the number of Issues that reach the repository, the language clearly needs to improve, because stability and robustness are needed to bring good projects to production
- Simplifying the syntax that could be used to write algorithms is possible because functional programming is minimalist.
- Improve the concurrency capabilities that consequently brings better performance in the computational capabilities of the DartVM because the syntax that allows functional programming allows us to simplify the way of processing large volumes of information in parallel, this is usable by algorithms that use GPU or NPUs that can create thousands of threads, these types of hardware accelerated solutions are the future
- Improve the satisfaction and morale of programmers because with minimal amounts of code a lot of work can be done, we must return to the time of minimalist philosophy that was very popular between 2013 and 2019, there are thousands of conferences on YouTube talking about this

At this point we should already understand the benefits of functional programming and why all competing languages ​​are adopting this paradigm, and Dart cannot be the exception. To do so, we must introduce the following new features:

| Name                                                                                                                         |  Status                   |
| ------------------------------------------------------------------------------- | ------------------ |
| Type inference                                                                                                          |    ✅                        |
| Immutability                                                                                                              |    ✅                        |
| Operator Overload                                                                                                    |    ✅                        |
| Higher Order Function                                                                                              |    ✅                        |
| [Functional Interfaces ](https://github.com/dart-lang/language/issues/3889)      |    ❌                        |
| [Optimize the DateTime type](https://github.com/dart-lang/language/issues/3910) |    ❌                        |
|  Immutable data structure                                                                                        |    ✅                         |
|  Async/Await                                                                                                              |    ✅                        |
|  Stream                                                                                                                      |    ⚠️                        |
|  [Create the type Monad](https://github.com/dart-lang/language/issues/3914)    |    ❌                        |
|  [Create the type Optional](https://github.com/dart-lang/language/issues/3731)    |    ❌                        |
|  Create the [dart:time](https://github.com/dart-lang/language/issues/3911) api    |    ❌                        |
|  Create the [dart:function](https://github.com/dart-lang/language/issues/3909) api  |    ❓                         |
|  Optimize the [dart:isolate](https://github.com/dart-lang/language/issues/3912) api |    ❌                        |
|  Optimize the [dart:async](https://github.com/dart-lang/language/issues/3913) api |    ⚠️                        |
|  Optimize the `dart:io` API to use functional programming                                    |    ⚠️                        |

To understand what emojis mean, see the following table

| Emoji               |  Meaning                                        |
| -------------- | --------------------------------- |
|    ✅                 | Implemented                                  |
|    ⚠️                 | Partially implemented                    |
|    ❓                 | Is not implemented                        |
|    ❌                 | Refused                                          |

**Comment by mraleph** [comment]
This issue is too grandiose and broad to be actionable. 

@Ing-Brayan-Martinez I am going to exercise my judgement here and politely request you to refrain from wasting team time with such issues. You are a repeatedly done this - in the last few months I already had to close multiple issues in this style filed by you. Please stop doing that. If you have some rather concrete issues with Dart e.g. bugs or feature requests backed by reasonable concrete development scenarios --- you are free to file them. However, in the future I ask you to refrain from filing grandiose _shift of paradigm_ style of issues. They are not helping anybody and members of the language team are simply wasting valuable time engaging on them.

Thank you for understanding.

**Comment by Ing-Brayan-Martinez** [comment]
@mraleph You're right, I'm wasting time, this time I wanted to do it clearly, with achievable goals because times have changed and the team has been reduced along with its budget, which is why I have divided it into several problems so as not to avoid exceeding the team's current capabilities.

The idea of ​​all these problems is to do it step by step in the proposed order and release them in different versions in a period of 3 to 4 years in the future. This is a long-term investment. I did not expect this to come out immediately.

Seeing the scenario, I am going to invest my time in the JVM using Kotlin, I will trust Dart again when one day they implement these new features, greetings and grateful

You have everything at your fingertips you have the power
